src/pool/pool_tbb.c

```diff
@@ -0,0 +1,238 @@
+
+    pthread_once(&tbb_is_initialized, load_tbb_symbols);
+
+    struct tbb_memory_pool *pool_data = malloc(sizeof(struct tbb_memory_pool));
+    if (!pool_data) {
+        fprintf(stderr, "cannot allocate memory for metadata\n");
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    pool_data->mem_provider = provider;
+    g_tbb_ops.pool_create_v1((intptr_t)pool_data, &policy,
+                             &(pool_data->tbb_pool));
+    *pool = (void *)pool_data;
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static void tbb_pool_finalize(void *pool) {
+    pthread_once(&tbb_is_initialized, load_tbb_symbols);
+    struct tbb_memory_pool *pool_data = (struct tbb_memory_pool *)pool;
+    g_tbb_ops.pool_destroy(pool_data->tbb_pool);
+}
+
+static void *tbb_malloc(void *pool, size_t size) {
+    struct tbb_memory_pool *pool_data = (struct tbb_memory_pool *)pool;
+    TLS_last_allocation_error = UMF_RESULT_SUCCESS;
+    void *ptr = g_tbb_ops.pool_malloc(pool_data->tbb_pool, size);
+    if (ptr == NULL) {
+        if (TLS_last_allocation_error == UMF_RESULT_SUCCESS) {
+            TLS_last_allocation_error = UMF_RESULT_ERROR_UNKNOWN;
+        }
+        return NULL;
+    }
+    return ptr;
+}
+
+static void *tbb_calloc(void *pool, size_t num, size_t size) {
+    assert(pool);
+    size_t csize = num * size;
+    void *ptr = tbb_malloc(pool, csize);
+    if (ptr == NULL) {
+        // TLS_last_allocation_error is set by tbb_malloc()
+        return NULL;
+    }
+
+    memset(ptr, 0, csize); // TODO: device memory is not accessible by host
+    return ptr;
+}
+
+static void *tbb_realloc(void *pool, void *ptr, size_t size) {
+    struct tbb_memory_pool *pool_data = (struct tbb_memory_pool *)pool;
+    TLS_last_allocation_error = UMF_RESULT_SUCCESS;
+    void *new_ptr = g_tbb_ops.pool_realloc(pool_data->tbb_pool, ptr, size);
+    if (new_ptr == NULL) {
+        if (TLS_last_allocation_error == UMF_RESULT_SUCCESS) {
+            TLS_last_allocation_error = UMF_RESULT_ERROR_UNKNOWN;
+        }
+        return NULL;
+    }
+    return new_ptr;
+}
+
+static void *tbb_aligned_malloc(void *pool, size_t size, size_t alignment) {
+    struct tbb_memory_pool *pool_data = (struct tbb_memory_pool *)pool;
+    TLS_last_allocation_error = UMF_RESULT_SUCCESS;
+    void *ptr =
+        g_tbb_ops.pool_aligned_malloc(pool_data->tbb_pool, size, alignment);
+    if (ptr == NULL) {
+        if (TLS_last_allocation_error == UMF_RESULT_SUCCESS) {
+            TLS_last_allocation_error = UMF_RESULT_ERROR_UNKNOWN;
+        }
+        return NULL;
+    }
+    return ptr;
+}
+
+static umf_result_t tbb_free(void *pool, void *ptr) {
+    struct tbb_memory_pool *pool_data = (struct tbb_memory_pool *)pool;
+    g_tbb_ops.pool_free(pool_data->tbb_pool, ptr);
+    return UMF_RESULT_SUCCESS;
+}
+
+static size_t tbb_malloc_usable_size(void *pool, void *ptr) {
+    (void)pool; // not used
+    (void)ptr;  // not used
+    return 0;   // unsupported
```

Comment:
I think there is pool_msize function in tbb. Can't we use it?

Reply:
Done